alibi/explainers/backends/tflow/cfrl_base.py

```diff
@@ -0,0 +1,523 @@
+                 predict_func: Callable,
+                 conditional_func: Callable,
+                 num_classes: int,
+                 batch_size: int,
+                 shuffle: bool = True) -> None:
+        """
+        Constructor.
+
+        Parameters
+        ----------
+        x
+            Array of input instances. The input should NOT be preprocessed as it will be preprocessed when calling
+            the `preprocessor` function.
+        preprocessor
+            Preprocessor function. This function correspond to the preprocessing steps applied to the autoencoder model.
+        predict_func
+            Prediction function. The classifier function should expect the input in the original format and preprocess
+            it internally in the `predict_func` if necessary.
+        conditional_func
+            Conditional function generator. Given an preprocesed input array, the functions generates a conditional
+            array.
+        num_classes
+            Number of classes in the dataset.
+        batch_size
+            Dimension of the batch used during training. The same batch size is used to infer the classification
+            labels of the input dataset.
+        shuffle
+            Whether to shuffle the dataset each epoch. `True` by default.
+        """
+        super().__init__()
+
+        self.x = x
+        self.preprocessor = preprocessor
+        self.predict_func = predict_func
+        self.conditional_func = conditional_func
+        self.num_classes = num_classes
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        # Infer the classification labels of the input dataset. This is performed in batches.
+        self.y_m = TfCounterfactualRLDataset.predict_batches(x=self.x,
+                                                             predict_func=self.predict_func,
+                                                             batch_size=self.batch_size)
+
+        # Preprocess data.
+        self.x = self.preprocessor(self.x)
+
+        # Generate shuffled indexes.
+        self.on_epoch_end()
+
+    def on_epoch_end(self) -> None:
+        """
+        This method is called every epoch and performs dataset shuffling.
+        """
+        self.indexes = np.arange(self.x.shape[0])
+
+        if self.shuffle:
+            np.random.shuffle(self.indexes)
+
+    def __len__(self) -> int:
+        return self.x.shape[0] // self.batch_size
+
+    def __getitem__(self, idx) -> Dict[str, np.ndarray]:
+        self.num_classes = np.clip(self.num_classes, a_min=0, a_max=2)  # TODO: remove this
+
+        # Select indices to be returned.
+        indexes = self.indexes[idx * self.batch_size:(idx + 1) * self.batch_size]
+
+        # Generate random target.
+        y_t = np.random.randint(low=0, high=self.num_classes, size=self.batch_size)
+
+        # compute conditional vector.
+        c = self.conditional_func(self.x[idx * self.batch_size: (idx + 1) * self.batch_size])
+
```

Comment:
There is no need here. In PyTorch, each worker processes a single item, and then a concatenation of the tensors is performed. Thus it is necessary to check if `c is not None` as a concatenation of `None`'s will raise an error. Since in TensorFlow we operate on batches, no concatenation is performed and thus no error in `c is None`.